### ISSUE TITLE
Is this Training Code? Pls review it!

### ISSUE THREAD

**Issue by spacewalkingninja** [vincent05r/FinCast-fts#12]
STEP BY STEP IMPLEMENTATION IN PSEUDO CODE

The following comprehensive PyTorch pseudo-code establishes the concrete implementation plan, transforming the theoretical interrogation and diagnostic resolutions into functional, production-ready modules.
Step 1: External PQ-Loss Formulation Module

This module addresses the critical absence of the internal loss calculation. It receives the massive `` prediction tensor, extracts the relevant vectors, calculates the Huber, Trend, and Pinball formulas precisely, and applies the boolean padding mask to ensure structural integrity.
Python
```

import torch
import torch.nn as nn
import torch.nn.functional as F

class PointQuantileLoss(nn.Module):
    def __init__(self, quantiles=[0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9], 
                 lambda_trend=0.1, lambda_quantile=1.0, huber_delta=0.1):
        super(PointQuantileLoss, self).__init__()
        self.quantiles = quantiles
        self.lambda_trend = lambda_trend
        self.lambda_quantile = lambda_quantile
        # Calibrated for Instance Normalized data
        self.huber_delta = huber_delta
        # reduction='none' allows application of the padding mask later
        self.huber_loss_fn = nn.HuberLoss(delta=self.huber_delta, reduction='none')

    def forward(self, predictions, targets, mask=None):
        """
        Executes external calculation of FinCast loss objectives.
        predictions shape: -> N=1 for univariate processing
        targets shape:
        mask shape: -> 1.0 for valid data, 0.0 for padding
        """
        # Isolate point forecast (mean) at index 0 and probabilstic boundaries
        pred_mean = predictions[:, :, :, 0]
        pred_quantiles = predictions[:, :, :, 1:]
        
        # 1. Compute Huber Point Loss
        loss_point = self.huber_loss_fn(pred_mean, targets)
        
        # 2. Compute Trend Consistency Loss via first-order differences
        # Execute differencing along the Horizon dimension (dim=2)
        diff_pred = pred_mean[:, :, 1:] - pred_mean[:, :, :-1]
        diff_target = targets[:, :, 1:] - targets[:, :, :-1]
        loss_trend_raw = (diff_pred - diff_target) ** 2
        
        # Re-pad the tensor to match original Horizon dimension for mask alignment
        pad_shape = (0, 1) # Pad 1 zero at the end of the H dimension
        loss_trend = F.pad(loss_trend_raw, pad_shape, "constant", 0.0)
        
        # 3. Compute Quantile Pinball Loss
        loss_quantile = torch.zeros_like(pred_mean)
        for i, q in enumerate(self.quantiles):
            q_pred = pred_quantiles[:, :, :, i]
            error = targets - q_pred
            # Apply asymmetric penalty based on percentile location
            pinball = torch.max(q * error, (q - 1) * error)
            loss_quantile += pinball
        loss_quantile = loss_quantile / len(self.quantiles)
        
        # Apply strict padding masks to prevent zero-value contamination
        if mask is not None:
            loss_point = loss_point * mask
            loss_trend = loss_trend * mask
            loss_quantile = loss_quantile * mask
            # Calculate active, non-padded elements (add epsilon to prevent div zero)
            active_elements = mask.sum() + 1e-8
        else:
            active_elements = targets.numel()
            
        # Aggregate and apply hyperparameters
        reduced_point = loss_point.sum() / active_elements
        reduced_trend = (loss_trend.sum() / active_elements) * self.lambda_trend
        reduced_quantile = (loss_quantile.sum() / active_elements) * self.lambda_quantile
        
        return reduced_point, reduced_trend, reduced_quantile
```

Step 2: Architecture Initialization and State Management

This module encapsulates the logic required to load the massive pre-trained checkpoint, systematically lock the required layers to optimize memory, execute the Monkey Patching injection, and explicitly expose the probabilistic output head for gradient updates.
Python
```

from peft_injector import wrap_with_peft
from models import PatchedTimeSeriesDecoder_MOE

def initialize_fincast_environment(checkpoint_path, device, learning_rate=1e-4):
    print("Initiating Base 1B Parameter Checkpoint Loading...")
    
    # Load foundational model directly into optimal precision format
    base_model = PatchedTimeSeriesDecoder_MOE.from_pretrained(checkpoint_path)
    base_model.to(device)
    
    # Execute total parameter freeze to protect pre-trained distribution mapping
    for param in base_model.parameters():
        param.requires_grad = False
        
    print("Executing PEFT Injection via Monkey Patching...")
    # Define targets to capture attention matrices and MoE expert projections
    target_modules = ["q_proj", "v_proj", "w1", "w2"] 
    # Wrap model, inserting low-rank adapters (r=8) with standard dropout
    model = wrap_with_peft(base_model, target_modules=target_modules, r=8, lora_dropout=0.05)
    
    print("Isolating and Unfreezing Quantile Output Head...")
    # Systematically search parameter names to locate the final projection matrix
    for name, param in model.named_parameters():
        if "output_projection" in name:  # Matches standard output head naming convention
            param.requires_grad = True
            
    # Diagnostic count to verify memory limits (~2M parameters expected)
    trainable_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
    print(f"Total Trainable Parameters Allocated: {trainable_params:,}")
    
    # Initialize AdamW optimizer, passing strictly the unfrozen parameters
    optimizer = torch.optim.AdamW(
        filter(lambda p: p.requires_grad, model.parameters()), 
        lr=learning_rate, 
        weight_decay=0.05 # Standard decay for time-series foundation models
    )
    
    return model, optimizer

```
Step 3: Orchestration of the Complete Training Loop

This function replaces the missing Training_DDP.py. It integrates the external loss module, implements gradient accumulation for VRAM safety, handles the staged learning rate schedule, and manages the execution of the validation loop to capture optimal weights.
Python
```

import torch
import math
from torch.optim.lr_scheduler import LambdaLR

def build_staged_lr_scheduler(optimizer, warmup_steps, plateau_steps, total_steps, min_lr_ratio=0.1):
    """
    Constructs the complex Warmup -> Plateau -> Cosine Decay trajectory.
    """
    def lr_lambda(current_step):
        if current_step < warmup_steps:
            # Linear Warmup phase
            return float(current_step) / float(max(1, warmup_steps))
        elif current_step < (warmup_steps + plateau_steps):
            # Stable Plateau phase
            return 1.0
        else:
            # Cosine Decay phase
            decay_steps = total_steps - warmup_steps - plateau_steps
            current_decay = current_step - warmup_steps - plateau_steps
            progress = float(current_decay) / float(max(1, decay_steps))
            return max(min_lr_ratio, 0.5 * (1.0 + math.cos(math.pi * progress)))
            
    return LambdaLR(optimizer, lr_lambda)

def execute_fincast_training_loop(model, train_loader, val_loader, optimizer, device, epochs=10, accum_steps=8):
    # Instantiate the custom loss calculation module
    pq_loss_fn = PointQuantileLoss(lambda_trend=0.1, lambda_quantile=1.0, huber_delta=0.1)
    
    # Calculate exact step counts based on batch size and accumulation rate
    total_opt_steps = len(train_loader) * epochs // accum_steps
    scheduler = build_staged_lr_scheduler(
        optimizer, 
        warmup_steps=int(total_opt_steps * 0.05),
        plateau_steps=int(total_opt_steps * 0.30),
        total_steps=total_opt_steps
    )
    
    # Deploy AMP GradScaler to manage float16 backward pass without underflow
    scaler = torch.cuda.amp.GradScaler()
    best_val_loss = float('inf')
    
    for epoch in range(epochs):
        model.train()
        optimizer.zero_grad()
        
        for step, batch in enumerate(train_loader):
            # Map batch tensors to hardware device
            x_context = batch['context'].to(device)
            y_target = batch['target'].to(device)
            freq_ids = batch['freq_map'].to(device) # Encodes the sub-daily resolution
            padding_mask = batch['target_mask'].to(device)
            
            # Execute forward pass strictly within mixed precision context
            with torch.cuda.amp.autocast(dtype=torch.float16):
                # The tokenization and instance normalization occur internally
                predictions, aux_moe_loss = model(inputs=x_context, freq_ids=freq_ids)
                
                # Divert tensors to external module to compute operational losses
                l_pt, l_tr, l_qt = pq_loss_fn(predictions, y_target, padding_mask)
                
                # Scale the MoE regularization loss to prevent router collapse
                scaled_moe_loss = 0.01 * aux_moe_loss 
                
                # Aggregate total loss and divide by accumulation steps to normalize gradients
                total_loss = (l_pt + l_tr + l_qt + scaled_moe_loss) / accum_steps
                
            # Execute backward pass via GradScaler
            scaler.scale(total_loss).backward()
            
            # Process gradient update only when accumulation threshold is met
            if (step + 1) % accum_steps == 0:
                # Unscale gradients to allow accurate mathematical clipping
                scaler.unscale_(optimizer)
                torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
                
                # Step optimizer and update scale multiplier
                scaler.step(optimizer)
                scaler.update()
                optimizer.zero_grad()
                scheduler.step() # Advance learning rate trajectory
                
        # --- Validation Phase ---
        model.eval()
        total_epoch_val_loss = 0.0
        
        with torch.no_grad(): # Suspend gradient tracking entirely
            for v_batch in val_loader:
                v_ctx = v_batch['context'].to(device)
                v_tgt = v_batch['target'].to(device)
                v_frq = v_batch['freq_map'].to(device)
                v_msk = v_batch['target_mask'].to(device)
                
                with torch.cuda.amp.autocast():
                    v_pred, v_aux = model(inputs=v_ctx, freq_ids=v_frq)
                    v_pt, v_tr, v_qt = pq_loss_fn(v_pred, v_tgt, v_msk)
                    # Compute un-accumulated validation sum
                    v_loss = v_pt + v_tr + v_qt + (0.01 * v_aux)
                    total_epoch_val_loss += v_loss.item()
                    
        avg_val_loss = total_epoch_val_loss / len(val_loader)
        
        # Checkpoint Management
        if avg_val_loss < best_val_loss:
            best_val_loss = avg_val_loss
            # Extract and save strictly the parameters that contain gradients
            state_dict_to_save = {k: v for k, v in model.state_dict().items() if v.requires_grad}
            checkpoint_data = {
                'epoch': epoch,
                'model_state_dict': state_dict_to_save,
                'optimizer_state_dict': optimizer.state_dict(),
                'validation_loss': best_val_loss
            }
            torch.save(checkpoint_data, f"fincast_lora_optimal_weights.pt")
            
        # Force garbage collection to prevent cross-epoch VRAM fragmentation
        torch.cuda.empty_cache()

```

**Comment by spacewalkingninja** [comment]
Hi! Pls check this code line by line against your training code, step by step, be systematic. Make sure this code is exact as real training code. Thanks. Make this equal as training code originally developed for Fincast-fts. Thanks. Be detailed, step by step, systematic, make sure you think exactly as the authors of this repo.
